### PR TITLE
potential solution for roundtrip issue with microseconds (issue #187)

### LINKED ISSUE #187: Back and forth conversions fail with num2date and date2num
I have problems going back and forth with the num2date and date2num routines.

One problem is a millisecond problem:
> import cftime as cf
> import datetime as dt
> dt1 = dt.datetime(1810, 4, 24, 16, 15, 10)
> cf.num2date(cf.date2num(dt1, units='days since -4713-01-01 12:00'), units='days since -4713-01-01 12:00')
> -> Out: cftime.DatetimeGregorian(1810, 4, 24, 16, 15, 9, 999999)
> cf.num2date(cf.date2num(dt1, units='days since 0001-01-01 12:00'), units='days since 0001-01-01 12:00')
> -> Out: cftime.DatetimeGregorian(1810, 4, 24, 16, 15, 9, 999999)
> cf.num2date(cf.date2num(dt1, units='days since 1600-01-01 12:00'), units='days since 1600-01-01 12:00')
> -> Out: cftime.DatetimeGregorian(1810, 4, 24, 16, 15, 9, 999999)
> cf.num2date(cf.date2num(dt1, units='days since 1700-01-01 12:00'), units='days since 1700-01-01 12:00')
> -> Out: cftime.DatetimeGregorian(1810, 4, 24, 16, 15, 10, 0)

The other one is off by one week if I use Lotus dates (or Excel 1904 dates):
> dt2 = dt.datetime(1271, 3, 18, 19, 41, 34)
> cf.num2date(cf.date2num(dt2, units='days since 1904-01-01 00:00'), units='days since 1904-01-01 00:00')
> -> Out: cftime.DatetimeGregorian(1271, 3, 11, 19, 41, 34, 0) 

I am the one who did the pull request once: [Unidata/netcdf4-python#433](https://github.com/Unidata/netcdf4-python/pull/433), adding a little offset corresponding to about 45 milliseconds. So I if do the same here, I get the right seconds but not the right microseconds, of course:
> import numpy as np
> eps = np.finfo(np.float).eps
> dt1 = dt.datetime(1810, 4, 24, 16, 15, 10)
> j1 = cf.date2num(dt1, units='days since 0001-01-01 12:00')
> cf.num2date(j1, units='days since 0001-01-01 12:00')                  
> -> Out: cftime.DatetimeGregorian(1810, 4, 24, 16, 15, 9, 999999)
> j1 += np.abs(j1) * eps
> cf.num2date(j1, units='days since 0001-01-01 12:00')
> -> Out: cftime.DatetimeGregorian(1810, 4, 24, 16, 15, 10, 9)

Perhaps you could introduce the rounding of microseconds again, just as you did for [issue #78](https://github.com/Unidata/cftime/issues/78).
I have no idea about the missing week though.

I use cftime version 1.2.0 on Python 3.8.3 on macOS 10.15.5.

Kind regards
Matthias

### PR DESCRIPTION
**Pull request by jswhit**
closes #187 and #189

### CODE REVIEW

File: cftime/_cftime.pyx [hunk 1]
```diff
@@ -364,8 +364,18 @@ def cast_to_int(num):
             raise OverflowError('time values outside range of 64 bit signed integers')
         if isinstance(num, np.ma.core.MaskedArray):
             int_num = np.ma.masked_array(np.rint(num), dtype=np.int64)
+            # use ceil instead of rint if 1 microsec less than a second
+            # or floor if 1 microsec greater than a second (issue #187)
+            int_num = np.ma.where(int_num%10000000 == 1, \
```
Comment (spencerkclark) [inline cftime/_cftime.pyx:369]: These should be 10^6, not 10^7, correct?
```suggestion
            int_num = np.ma.where(int_num%1000000 == 1, \
```

File: cftime/_cftime.pyx [hunk 2]
```diff
@@ -364,8 +364,18 @@ def cast_to_int(num):
...
+                      np.ma.masked_array(np.floor(num),dtype=np.int64), int_num)
+            int_num = np.ma.where(int_num%10000000 == 9999999, \
```
Comment (spencerkclark) [inline cftime/_cftime.pyx:371]:
```suggestion
            int_num = np.ma.where(int_num%1000000 == 999999, \
```

File: cftime/_cftime.pyx [hunk 3]
```diff
@@ -364,8 +364,18 @@ def cast_to_int(num):
...
+                      np.ma.masked_array(np.ceil(num),dtype=np.int64), int_num)
         else:
             int_num = np.array(np.rint(num), dtype=np.int64)
+            int_num = np.where(int_num%10000000 == 1, \
```
Comment (spencerkclark) [inline cftime/_cftime.pyx:375]:
```suggestion
            int_num = np.where(int_num%1000000 == 1, \
```

File: cftime/_cftime.pyx [hunk 4]
```diff
@@ -364,8 +364,18 @@ def cast_to_int(num):
...
+                      np.array(np.floor(num),dtype=np.int64), int_num)
+            int_num = np.where(int_num%10000000 == 9999999, \
```
Comment (spencerkclark) [inline cftime/_cftime.pyx:377]:
```suggestion
            int_num = np.where(int_num%1000000 == 999999, \
```

File: cftime/_cftime.pyx (hunk 1 above)
Comment (jswhit) [reply cftime/_cftime.pyx]: Right - good catch.  I've also added a check on units, so this is only done if units are not microsecs or millisecs (where this kind of rounding is probably not desirable).

**Comment by jswhit** [comment]
Fix included for calculation of leap years when year is negative (issue #188). Closes issue #188.

File: Changelog [hunk 5]
```diff
@@ -3,6 +3,10 @@ version 1.2.1 (not yet released)
  * num2date uses 'proleptic_gregorian' scheme when basedate is post-Gregorian but date is pre-Gregorian
    (issue #182).
  * fix 1.2.0 regression (date2num no longer works with numpy scalar array inputs, issue #185).
+ * Fix for issue #187 (have date2num round to the nearest second when within 1
+   microsecond).
+ * Fix for issue #188 (leap years calculated incorrectly for negative years in
```
Comment (spencerkclark) [inline Changelog:8]:
```suggestion
 * Fix for issue #189 (leap years calculated incorrectly for negative years in
```

**Pull request merged**